app/code/Magento/Ui/view/base/web/js/dynamic-rows/dynamic-rows.js

```diff
@@ -548,6 +548,13 @@ define([
                     });
 
                     this.labels.push(data);
+
+                    /**
+                     * Sort the array after an element was added to fix an bug where
+                     * additional added field labels in ui_components haven't the right
+                     * sort order.
+                     */
+                    this.labels.sort(this._compare);
```

Comment:
Could you, please, move push and sort calls outside of loop for performance improvement. Also, I think it's better to use `_.sortBy()` function and remove custom compare function. An example:
```js
initHeader: function () {
    var labels = [],
        data;

    if (!this.labels().length) {
        _.each(this.childTemplate.children, function (cell) {
            data = this.createHeaderTemplate(cell.config);
            cell.config.labelVisible = false;
            _.extend(data, {
                label: cell.config.label,
                name: cell.name,
                required: !!cell.config.validation,
                columnsHeaderClasses: cell.config.columnsHeaderClasses,
                sortOrder: cell.config.sortOrder
            });

            labels.push(data);
        }, this);
        this.labels(_.sortBy(labels, 'sortOrder'));
    }
}
```